### PR TITLE
Added the octogone mission planner

### PR DESCRIPTION
**Pull request by Ayoubsab**

### CODE REVIEW

File: catkin_ws/src/planner/src/substates/octogone_task.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                print("No detection.... Timeout in " + str((startTime + timeout) - time.time()), end='\r')
+                if time.time() > startTime + timeout:
+                    print("Octogone no longer visible.")
+                    return 'failure'
+                    
+        print("Successfully centered octogone!")
+        return 'success'
+
+        
+class goThroughOctogone(smach.State):
```
Comment (frompotenza) [inline catkin_ws/src/planner/src/substates/octogone_task.py]: name of the class should start with a Capital letter

File: catkin_ws/src/planner/src/mission.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 #!/usr/bin/env python3
 
+from planner.src.substates.octogone_task import CenterOctogone, goThroughOctogone
```
Comment (antoinedang) [inline catkin_ws/src/planner/src/mission.py]: should be from src.substates.octogone_task import CenterOctogone, goThroughOctogone

File: catkin_ws/src/planner/src/mission.py [hunk 3]
```diff
@@ -68,6 +69,22 @@ def Tricks(t):
             res = "trick not identified"
     endMission("Finished trick. Result {}".format(res))
 
+
+def octogone_task():
+    sm = smach.StateMachine(outcomes=['success', 'failure'])
+    ocotogne_target_class = 3 
+    with sm:
+        smach.StateMachine.add('find_octogone', BreadthFirstSearch(timeout=9999, target_classes=[(ocotogne_target_class, 1)], control=control, mapping=mapping),
```
Comment (antoinedang) [inline catkin_ws/src/planner/src/mission.py:79]: We should move the AUV to the octagon approximate position before breadth first search

Comment (antoinedang) [reply catkin_ws/src/planner/src/mission.py:79]: ideally should also support having several octagon position and at each failure we try the next one

File: catkin_ws/src/planner/src/substates/octogone_task.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        z_increment = 1
+
+        startTime = time.time()
+
+        while not (centered):
+            if len(last_object_detection) > 0:
+                startTime = time.time()
+                center_x = last_object_detection[0]
+                center_y = last_object_detection[1]
+                last_object_detection = []
```
Comment (antoinedang) [inline catkin_ws/src/planner/src/substates/octogone_task.py]: doesn't use mapping, instead you should take mapping as an argument to this class, and use the methods supplied in src.substates.utility.mapping to navigate the gate -> you can assume if you've arrived at the navigation step that there is an octagon

File: catkin_ws/src/planner/src/substates/octogone_task.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+                print("Octogone in view at: x:{}, y:{}!".format(center_x, center_y))
+                
+                centering_error = math.sqrt((center_x - targetCenterX)**2 + (center_y - targetCenterY)**2)
+
+                if centering_error > centering_tolerance: #isn't centered
```
Comment (antoinedang) [inline catkin_ws/src/planner/src/substates/octogone_task.py]: once you use mapping you won't need this, you can just do control.move(octagon_x, octagon_y) to go above it

File: catkin_ws/src/planner/src/substates/octogone_task.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+        
+    def execute(self, ud):
+        '''
+        Moves the AUV to the surface (ie, z=0)
+        '''
+        
+
+        print("Starting movement through octogone.")
+        curr_x, curr_y, _ = self.state.getPosition()
+        self.control.move((curr_x, curr_y, -1))
```
Comment (antoinedang) [inline catkin_ws/src/planner/src/substates/octogone_task.py]: moving to your current position is redundant. You could just replace both the navigation and goThrough classes by this class and instead move to the octagon's x and y. the rest is correct

File: catkin_ws/src/planner/src/substates/octogone_task.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        self.control.kill()
+
+        return 'succes'
```
Comment (antoinedang) [inline catkin_ws/src/planner/src/substates/octogone_task.py]: 'success'

**Review by antoinedang — CHANGES_REQUESTED** [review]
see comments

**Pull request closed without merging**